Mathlib/LinearAlgebra/Reflection.lean

```diff
@@ -0,0 +1,161 @@
+
+## Main definitions / results:
+
+ * `Module.preReflection`: the definition of the map `y ↦ y - (f y) • x`. Its main utility lies in
+   the fact that it does not require the assumption `f x = 2`, giving the user freedom to defer
+   discharging this proof obligation.
+ * `Module.reflection`: the definition of the map `y ↦ y - (f y) • x`. This requires the assumption
+   that `f x = 2` but by way of compensation it produces a linear equivalence rather than a mere
+   linear map.
+ * `Module.Dual.eq_of_preReflection_image_subset`: a uniqueness result about reflections preserving
+   finite spanning sets that is useful in the theory of root data / systems.
+
+## TODO
+
+Related definitions of reflection exists elsewhere in the library. These more specialised
+definitions, which require an ambient `InnerProductSpace` structure, are `reflection` (of type
+`LinearIsometryEquiv`) and `EuclideanGeometry.reflection` (of type `AffineIsometryEquiv`). We
+should connect (or unify) these definitions with `Module.reflecton` defined here.
+
+-/
+
+open Set Function Pointwise
+open Module hiding Finite
+open Submodule (span)
+
+noncomputable section
+
+variable {R M : Type*} [CommRing R] [AddCommGroup M] [Module R M] (x y : M) (f : Dual R M)
+
+namespace Module
+
+/-- Given an element `x` in a module `M` and a linear form `f` on `M`, we define the endomorphism
+of `M` for which `y ↦ y - (f y) • x`.
+
+One is typically interested in this endomorphism when `f x = 2`; this definition exists to allow the
+user defer discharging this proof obligation. See also `Module.reflection`. -/
+def preReflection : End R M :=
+  LinearMap.id - f.smulRight x
+
+lemma preReflection_apply :
+    preReflection x f y = y - (f y) • x := by
+  simp [preReflection]
+
+variable {x f}
+
+lemma preReflection_apply_self (h : f x = 2) :
+    preReflection x f x = - x := by
+  rw [preReflection_apply, h, two_smul]; abel
+
+lemma involutive_preReflection (h : f x = 2) :
+    Involutive (preReflection x f) :=
+  fun y ↦ by simp [h, smul_sub, two_smul, preReflection_apply]
+
+lemma preReflection_preReflection (g : Dual R M) (h : f x = 2) :
+    preReflection (preReflection x f y) (preReflection f (Dual.eval R M x) g) =
+    (preReflection x f) ∘ₗ (preReflection y g) ∘ₗ (preReflection x f) := by
+  ext m
+  simp only [h, preReflection_apply, mul_comm (g x) (f m), mul_two, mul_assoc, Dual.eval_apply,
+    LinearMap.sub_apply, LinearMap.coe_comp, LinearMap.smul_apply, smul_eq_mul, smul_sub, sub_smul,
+    smul_smul, sub_mul, comp_apply, map_sub, map_smul, add_smul]
+  abel
+
+/-- Given an element `x` in a module `M` and a linear form `f` on `M` for which `f x = 2`, we define
+the endomorphism of `M` for which `y ↦ y - (f y) • x`.
+
+It is an involutive endomorphism of `M` fixing the kernel of `f` for which `x ↦ -x`. -/
+def reflection (h : f x = 2) : M ≃ₗ[R] M :=
+  LinearEquiv.ofBijective (preReflection x f) (involutive_preReflection h).bijective
```

Comment:
I guess you can get better defeqs if you explicitly give the inverse function. There might even be an `involutive.toEquiv` or something like that, so that it will still be a one-liner.

Reply:
Oh, thanks `Function.Involutive.toPerm` is just what I needed.